src/function/cast/variant/to_variant.cpp

```diff
@@ -190,4 +191,216 @@ BoundCastInfo DefaultCasts::ImplicitToVariantCast(BindCastInput &input, const Lo
+	}
+
+	unsafe_vector<idx_t> non_null_child_counts(count, 0);
+	ValidityMask key_is_used(children.size());
+	key_is_used.SetAllInvalid(children.size());
+	for (idx_t child_idx = 0; child_idx < children.size(); child_idx++) {
+		const auto &child_format = child_source_data[child_idx].source_format;
+		for (idx_t i = 0; i < count; i++) {
+			if (!source_validity.RowIsValid(source[i])) {
+				continue;
+			}
+			if (child_format.validity.RowIsValid(child_format.sel->get_index(i))) {
+				non_null_child_counts[i]++;
+				key_is_used.SetValidUnsafe(child_idx);
+			}
+		}
+	}
+
+	//! Look up all the dictionary indices for the struct keys
+	vector<uint32_t> dictionary_indices;
+	dictionary_indices.reserve(children.size());
+
+	auto &variant = result.variant;
+	ContainerSelectionVectors sel(count);
+	for (idx_t i = 0; i < count; i++) {
+		const auto index = source[i];
+		const auto result_index = selvec ? selvec->get_index(i) : i;
+
+		auto &blob_offset = blob_offset_data[result_index];
+
+		const auto &non_null_child_count = non_null_child_counts[i];
+		if (source_validity.RowIsValid(index) && non_null_child_count != 0) {
+			WriteVariantMetadata<WRITE_DATA>(result, result_index, values_offset_data, blob_offset, values_index_selvec,
+			                                 i, VariantLogicalType::OBJECT);
+			WriteContainerData<WRITE_DATA>(result.variant, result_index, blob_offset, non_null_child_count,
+			                               children_offset_data[result_index]);
+
+			if (WRITE_DATA && dictionary_indices.empty()) {
+				if (StructType::IsUnnamed(type)) {
+					throw ConversionException("Can't cast unnamed struct to VARIANT");
+				}
+				auto &struct_children = StructType::GetChildTypes(type);
+				for (idx_t child_idx = 0; child_idx < children.size(); child_idx++) {
+					if (key_is_used.RowIsValidUnsafe(child_idx)) {
+						auto &struct_child = struct_children[child_idx];
+						string_t struct_child_str(struct_child.first.c_str(),
+						                          NumericCast<uint32_t>(struct_child.first.size()));
+						dictionary_indices.push_back(result.GetOrCreateIndex(struct_child_str));
+					} else {
+						dictionary_indices.push_back(static_cast<uint32_t>(-1));
+					}
+				}
+			}
+
+			//! children
+			if (WRITE_DATA) {
+				auto &children_list_entry = variant.children_data[result_index];
+				auto &keys_list_entry = variant.keys_data[result_index];
+
+				idx_t children_index = children_list_entry.offset + children_offset_data[result_index];
+				idx_t keys_offset = keys_list_entry.offset + keys_offset_data[result_index];
+
+				idx_t non_null_idx = 0; // Keep track of non-NULL children because this is a sparse conversion
+				for (idx_t child_idx = 0; child_idx < children.size(); child_idx++) {
+					const auto &child_format = child_source_data[child_idx].source_format;
+					if (!child_format.validity.RowIsValid(child_format.sel->get_index(i))) {
+						continue; // Sparse conversion: skip NULL
```

Comment:
Looking at this, this seems like the only change made to the function - where the rest is copy pasted from `ConvertStructToVariant`.

If we decide to keep this function after all - I think it would be better to add a template parameter to `ConvertToVariant ` to enable this behavior, i.e.:


```cpp
template <bool WRITE_DATA, bool IGNORE_NULLS, bool SPARSE_CONVERSION = false>
 ConvertToVariant(...)
```

And then call that here:

```cpp
ConvertToVariant<...., true>(...)
```

Copy pasting large amounts of logic is not great to begin with - but here I think it's especially risky as this function is (1) very complex, and (2) new and not well tested.